File: linear_scan.py
import os
import time
import argparse
import random
import numpy as np

_POPCOUNT_TABLE = np.array([bin(i).count("1") for i in range(256)], dtype=np.uint8)


def hamming_distances(codes: np.ndarray, q: np.ndarray) -> np.ndarray:
    """
    codes : (N, D) uint8
    q     : (D,)   uint8
    returns (N,) int32 Hamming distances
    """
    xor = np.bitwise_xor(codes, q[np.newaxis, :])
    return _POPCOUNT_TABLE[xor].sum(axis=1).astype(np.int32)


def linear_scan_range(codes: np.ndarray, q: np.ndarray, r: int) -> np.ndarray:
    """Return indices of rows in codes whose Hamming distance to q is <= r."""
    d = hamming_distances(codes, q)
    return np.flatnonzero(d <= r).astype(np.int32)

def run_linear_scan(dataset_dir: str, dataset: str, radius: int, query_count: int, seed: int = None) -> None:
    from dataset import read_fvecs, binarize

    if seed is None:
        seed = random.randint(0, 2**32 - 1)

    base_path  = os.path.join(dataset_dir, f'{dataset}_base.fvecs')
    query_path = os.path.join(dataset_dir, f'{dataset}_query.fvecs')

    if not os.path.exists(base_path):
        raise FileNotFoundError(f"Dataset not found: {base_path}")

    print(f"Loading {dataset} ...")
    base    = read_fvecs(base_path)
    queries = read_fvecs(query_path)
    print(f"  base    : {base.shape}")
    print(f"  queries : {queries.shape}")

    print(f"Binarizing  (GRP seed={seed}) ...")
    base_codes  = binarize(base, seed=seed)
    query_codes = binarize(queries, seed=seed)
    print(f"  code shape : {base_codes.shape}  dtype={base_codes.dtype}")

    Q = min(query_count, len(query_codes))
    print(f"\nLinear scan  r={radius}, Q={Q} ...")

    results = []
    running_avg = 0.0
    for i in range(Q):
        t0 = time.perf_counter()
        res = linear_scan_range(base_codes, query_codes[i], radius)
        qt = time.perf_counter() - t0
        running_avg += (qt - running_avg) / (i + 1)
        results.append(res)

    total    = running_avg * Q
    avg_hits = float(np.mean([len(r) for r in results]))

    print(f"  total        : {total:.3f} s")
    print(f"  avg query    : {running_avg * 1000:.3f} ms  (running average)")
    print(f"  QPS          : {Q / total:.1f}")
    print(f"  avg hits     : {avg_hits:.1f}")


def main() -> None:
    from dataset import datasets, datasets_dir

    ap = argparse.ArgumentParser(description='Linear scan Hamming range query benchmark')
    ap.add_argument('--dataset', choices=list(datasets), default='sift',
                    help='dataset to use (default: sift)')
    ap.add_argument('--radius', type=int, default=16,
                    help='Hamming radius for range query (default: 16)')
    ap.add_argument('--query_count', type=int, default=100,
                    help='number of queries to run (default: 100)')
    ap.add_argument('--seed', type=int, default=None,
                    help='RNG seed for GRP projection matrix (default: random)')
    args = ap.parse_args()

    dataset      = datasets[args.dataset]
    dataset_dir = os.path.join(datasets_dir, dataset)
    run_linear_scan(dataset_dir, dataset, args.radius, args.query_count, args.seed)


if __name__ == '__main__':
    main()

File: cpp/linear_scan.cpp
// Linear-scan Hamming range query, C++ version.
//
// Mirrors linear_scan.py: load fvecs, binarize with GRP, run Q queries,
// report per-query and average timing.
//
// Usage:
//   ./linear_scan --dataset siftsmall --radius 20 --query_count 10 --seed 0
#include <cstdio>
#include <iostream>
#include <vector>

#include "common.hpp"

namespace fs = std::filesystem;
using common::ByteMatrix;
using common::FloatMatrix;

// Return indices i with hamming(codes[i], q) <= r.
static std::vector<int32_t> linear_scan_range(const ByteMatrix& codes,
                                              const uint8_t* q,
                                              int r)
{
    std::vector<int32_t> hits;
    hits.reserve(64);
    const std::size_t N = codes.n;
    const std::size_t D = codes.D;
    for (std::size_t i = 0; i < N; ++i) {
        if (common::hamming(codes.row(i), q, D) <= r) {
            hits.push_back(static_cast<int32_t>(i));
        }
    }
    return hits;
}

int main(int argc, char** argv)
{
    common::Args args(argc, argv);
    const std::string dataset     = args.get("dataset", "siftsmall");
    const int         radius      = args.get_int("radius", 20);
    const int         query_count = args.get_int("query_count", 10);
    const int         seed        = args.get_int("seed", 0);

    const fs::path ddir = common::dataset_dir(dataset);
    const std::string base_path  = (ddir / (dataset + "_base.fvecs")).string();
    const std::string query_path = (ddir / (dataset + "_query.fvecs")).string();

    std::printf("Loading %s ...\n", dataset.c_str());
    common::Timer tload;
    FloatMatrix base    = common::read_fvecs(base_path);
    FloatMatrix queries = common::read_fvecs(query_path);
    std::printf("  base    : (%zu, %zu)\n", base.n, base.d);
    std::printf("  queries : (%zu, %zu)\n", queries.n, queries.d);
    std::printf("  load    : %.2f s\n", tload.s());

    std::printf("Binarizing (GRP seed=%d) ...\n", seed);
    common::Timer tbin;
    ByteMatrix base_codes  = common::binarize(base,    static_cast<uint32_t>(seed));
    ByteMatrix query_codes = common::binarize(queries, static_cast<uint32_t>(seed));
    std::printf("  packed base    : (%zu, %zu) bytes  bits=%zu\n",
                base_codes.n, base_codes.D, base_codes.bits());
    std::printf("  binarize       : %.2f s\n", tbin.s());

    const int Q = std::min<int>(query_count, static_cast<int>(query_codes.n));
    std::printf("\nLinear scan  r=%d, Q=%d ...\n", radius, Q);

    double running_avg_ms = 0.0;
    long long total_hits = 0;
    for (int i = 0; i < Q; ++i) {
        common::Timer tq;
        auto hits = linear_scan_range(base_codes, query_codes.row(i), radius);
        double qms = tq.ms();
        running_avg_ms += (qms - running_avg_ms) / (i + 1);
        total_hits += static_cast<long long>(hits.size());
        std::printf("  q[%02d]  scan=%7.2f ms  hits=%6zu\n", i, qms, hits.size());
    }

    double total_ms = running_avg_ms * Q;
    double avg_hits = static_cast<double>(total_hits) / Q;

    std::printf("\n  total        : %.3f s\n",        total_ms * 1e-3);
    std::printf("  avg query    : %.3f ms\n",         running_avg_ms);
    std::printf("  QPS          : %.1f\n",            Q / (total_ms * 1e-3));
    std::printf("  avg hits     : %.1f\n",            avg_hits);
    return 0;
}
